<h1 style="font-size:42px; text-align:center; margin-bottom:30px;"><span style="color:SteelBlue">Module 2:</span> Python Crash Course</h1>
<hr>

Welcome to the **Companion Workbook** for <span style="color:royalblue">Module 2: Python Crash Course</span> (and for the <span style="color:royalblue">Expanded Python Crash Course</span>)!

After you read each lesson, come back here to practice any lesson code and complete the accompanying exercises.
* You can insert "scratch" code cells using **Insert > Insert Cell Below/Above** in the toolbar.
* Remember, to run a cell, you can click into it anywhere and press <code style="color:steelblue">Shift + Enter</code>.
* Use the table of contents below to find the lesson you're working on.
* Any "mission-critical" lesson code required for the exercises (such as importing libraries or data files) is included as well.

**Tip:** There are multiple ways to solve each exercise... these solutions are only one way!

<br><hr id="toc">

### Table of Contents

* [Lesson 1: Python Basics](#l1)
* [Lesson 2: Data Structures](#l2)
* [Lesson 3: Flow and Functions](#l3)
* [Lesson 4: NumPy](#l4)
* [Lesson 5: Pandas](#l5)

<br><hr>

## A few quick notes before we begin

Before we do anything else, let's import the <code style="color:steelblue">print()</code> function from the "future" to ensure our code is compatible with Python 3.
* Remember, you can run code cells by clicking anywhere in the cell to highlight it, then pressing <code style="color:steelblue">Shift + Enter</code> on your keyboard.

In [ ]:
from __future__ import print_function  # Compatability with Python 3
print('Print function ready to serve.')

When running **multiple lines of code** in Jupyter Notebooks, only the last result is shown.

In [ ]:
3 * 5  # Not shown
16 / 4 # Shown

But you can display each line by explicitly printing them with the <code style="color:steelblue">print()</code> function.

In [ ]:
print( 3 * 5 ) # Shown
print( 16 / 4 ) # Shown

Hey, did you see that <code style="color:dimgray; font-weight:bold">Gray</code> text in the code cell? 

That's called a **comment**.
* Comments add extra information, and they are not executed
* In Python, comments start with the pound sign (a.k.a. hashtag): 

<pre style="color:dimgray"># This is a comment</pre>

In [ ]:
# print( 3 * 5 )  <-- This code does not get run

By the way, if you <code style="color:steelblue">print()</code> multiple objects, separated by commas, it will **concatenate** those objects into a single string (more on strings later).

In [10]:
# Print concatenation
print( 'Testing',',1,2,4' )

Testing ,1,2,4


Great, now let's start our tour of the basics of the Python programming language.

By the way, you'll  practice each of these topics throughout the course, so don't worry too much about remembering every little detail on your first pass.

<br id="l1">

# Lesson 1: Python Basics

First, import the print function (if you've closed and reopened Jupyter since last time).

In [ ]:
from __future__ import print_function  # Compatability with Python 3
print('Print function ready to serve.')

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">
## <span style="color:RoyalBlue">Exercise 1.2</span>

It's time for your first exercise!

This is actually the start of a long series of connected exercises. These will all belong to the same **training mission**, and they will continue through the entire Python Crash Course.

***Here's the setting...***

> *As a new recruit, you're being sent on a training mission.*

> * *You will travel to different locations in London.*
> * *First, you will prepare for and plan the trip.*
> * *After you arrive at each location, you will meet with local business owners and offer your help.*

<p style="text-align:center; margin-top:20px; margin-bottom:20px;">
<img src="https://elitedatascience.com/wp-content/uploads/2017/02/caseys-flower-shop.jpg" alt="Casey's Flower Shop" style="width:280px; box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);">
<br>
<em>You have a strange feeling that you'll see this banner again...</em>
</p>

> *Simple enough, right? Let's get started.*

<br>

**First, calculate the total volume of water you can carry, in cm$^3$.**
* The formula for the volume of a cylinder is $V = \pi r^2 h$

In [17]:
# Total volume of water you can carry
import math as m
n_bottles = 3
radius = 4
height = 16
pi = m.pi
bottle_volume = pi*height*radius**2
total_volume = bottle_volume*3
print(total_volume)

2412.743157956961


**Q estimates that you'll need at least 2000 cm$^3$ for the mission. Print <code style="color:steelblue">True</code> if you have enough space or <code style="color:steelblue">False</code> if you do not.**
* Use the **greater-than-or-equal-to** operator.

In [18]:
# Do you have space for at least 2000 cm^3
print(total_volume >= 2000)

True


**If Q has 5000 cm$^3$ of water, what's the maximum number of bottles can you fill completely?**
* The output should be an integer (completely full bottles)

In [26]:
print(m.floor(5000/bottle_volume))

6


<hr style="border-color:royalblue;background-color:royalblue;height:1px;">
## <span style="color:RoyalBlue">Exercise 1.3</span>

You already told Q that you have enough space for water, but he didn't hear you clearly... You might need to just print him a message.

<br>
**Print the message:**

<pre>Yes, I have enough space. I can carry <span style="color:steelblue">total_volume</span> cm^3 of water</pre>

* Dynamically insert value you calculated in <span style="color:royalblue">Exercise 1.2</span> in place of <code style="color:steelblue">total_volume</code>.

In [22]:
# Print message to Q
print('Yes, I have enough space. I can carry {} cm^3 of water'.format(total_volume))

Yes, I have enough space. I can carry 2412.743157956961 cm^3 of water


<hr style="border-color:royalblue;background-color:royalblue;height:1px;">
## <span style="color:RoyalBlue">Exercise 1.4</span>

Repeat the calculations from <span style="color:RoyalBlue">Exercise 1.2</span>, but this time use variables with descriptive names.

* You have **3** empty cylinder-shaped bottles. 
* Each bottle has a height of **16** cm and a radius of **4** cm.
* Each bottle can be completely filled (ignore the thickness of the bottle).
* The formula for the volume of a cylinder is $V = \pi r^2 h$

<br>
**Start by setting variables with descriptive names for bottle dimensions, the number of bottles, and $\pi$.**

In [23]:
# Set variables
import math as m
n_bottles = 3
radius = 4
height = 16
pi = m.pi
bottle_volume = pi*height*radius**2
total_volume = bottle_volume*3
print(total_volume)

2412.743157956961


<br>
**Next, calculate the intermediary step of a single bottle's volume.**
* Set it to a new variable.

In [24]:
# Volume of one bottle (in cm^3)
bottle_volume = pi*height*radius**2
# Print volume of a single bottle
print(bottle_volume)

804.247719318987


<br>
**Finally, calculate total volume.**
* Set it to a new variable.

In [25]:
# Total volume
total_volume = bottle_volume*3
# Print total volume
print(total_volume)


2412.743157956961


<br>

***The mission continues in the next lesson...***

> *Now that you've stocked up on enough water, you're ready to go on your training mission into London.*

> *In the next lesson, we'll start planning the trip and all the locations you'll visit.*

<br>
<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

<div style="text-align:center; margin: 40px 0 40px 0;">
[**Back to Contents**](#toc)
</div>

<br id="l2">
# Lesson 2: Data Structures

First, import the print function (if you've closed and reopened Jupyter since last time).

In [ ]:
from __future__ import print_function  # Compatability with Python 3
print('Print function ready to serve.')

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">
## <span style="color:RoyalBlue">Exercise 2.1</span>

***In the previous lesson...***

> *In the previous lesson, you learned about the training mission: a visit to London.*

> *You also made sure you had enough water for the trip by performing calculations.*


Now, it's time to start planning the locations in London to visit!

<p style="text-align:center; margin-top:30px; margin-bottom:30px">
<img src="https://elitedatascience.com/wp-content/uploads/2017/02/london-map-boroughs.png" alt="London Boroughs Map" width=640 style="box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);">
<br>
<em>Boroughs of London.</em>
</p>

Before we start, let's import lists of locations by borough. 
* These are stored in text files that Q has provided for you. 
* Python has a variety of **input/output** methods. We won't cover them here, but you can learn more about them in the [documentation](https://docs.python.org/2/tutorial/inputoutput.html).

<br>
**First, run this code.**

In [6]:
# Read lists of locations (simply run this code block)
with open('project_files/brent.txt', 'r') as f:
    brent = f.read().splitlines()
    with open('project_files/camden.txt', 'r') as f:
    camden = f.read().splitlines()

with open('project_files/redbridge.txt', 'r') as f:
    redbridge = f.read().splitlines()

with open('project_files/southwark.txt', 'r') as f:
    southwark = f.read().splitlines()

Note that when text files are read using the <code style="color:steelblue">splitlines()</code> function, the resulting object is a list.

So the four objects you just created from the files - <code style="color:steelblue">brent</code>, <code style="color:steelblue">camden</code>, <code style="color:steelblue">redbridge</code>, and <code style="color:steelblue">southwark</code> - are all lists.

In [31]:
print(type(brent))

<class 'list'>


Let's start exploring this data.

<br>
**Print the first 5 locations in Camden.**

In [33]:
# Print the first 5 locations in Camden
print(camden[0:5])

['Chalk Farm', 'Dartmouth Park', 'Fitzrovia', 'Frognal', 'Bloomsbury']


Next, we need to know how many locations are in each borough. 

<br>
**Print the number of locations in each list.**
* Which borough has the most locations?

In [34]:
# Print length of each list
print('Brent: ', len(brent))
print('Camden: ', len(camden))
print('Redbridge: ', len(redbridge))
print('Southwark: ', len(southwark))


Brent:  19
Camden:  22
Redbridge:  18
Southwark:  15


Next, Q wants you to answer a couple of questions... His memory is not as good as it used to be. 

He asks you to:
* **Print <code style="color:steelblue">True</code> if <code>'Newbury Park'</code> is in Redbridge or <code style="color:steelblue">False</code> if it's not.**
* **Print <code style="color:steelblue">True</code> if <code>'Peckham'</code> is in Brent or <code style="color:steelblue">False</code> if it's not.**

In [40]:
# Is 'Newbury Park' in Redbridge?

print('Newbury Park' in(redbridge))
# Is 'Peckham' in Brent?
print('Peckham' in(brent))


True
False


**Finally, print the minimum and maximum values in the Southwark list.**
* Why do you think you get those results?
* How do you think lists of strings are ordered?

In [42]:
# Print minimum value in southwark
print(min(southwark))

# Print maximum value in southwark
print( max(southwark))


Bankside
Walworth


<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 2.3</span>
Let's continue planning locations to visit for our training mission to London. Before we continue, we need to remove duplicates from our lists because we don't have time to visit the same location twice.

<br>
**For each of the 4 lists of locations, print <code style="color:steelblue">True</code> if it has duplicate locations and <code style="color:steelblue">False</code> if it doesn't.**
* Hint: A list with duplicates will have a greater length than a set of the same locations.

In [44]:
# Brent has duplicates?
print(len(brent) > len(set(brent)))

# Camden has duplicates?
print(len(camden) > len(set(camden)))

# Redbridge has duplicates?
print(len(redbridge) > len(set(redbridge)))

# Southwark has duplicates?
print(len(southwark) > len(set(southwark)))

False
True
True
True


<br>
**For the lists with duplicates, remove duplicates by converting them into sets. Then, convert them back into lists.**
* Hint: <code style="color:steelblue">set()</code> and <code style="color:steelblue">list()</code> are your friends.

In [54]:
# Convert lists to sets to remove duplicates, then convert them back to lists 
camden = list(set(camden))
redbridge = list(set(redbridge))
southwark = list(set(southwark))


['Bloomsbury',
 "King's Cross",
 'Primrose Hill',
 'Kentish Town',
 'Holborn',
 'Hampstead',
 'Dartmouth Park',
 'Highgate',
 'West Hampstead',
 'Camden Town',
 'Fitzrovia',
 'Frognal',
 'Swiss Cottage',
 'Chalk Farm',
 'Belsize Park',
 'Gospel Oak',
 'St Giles',
 'Somerstown',
 'St Pancras']

Great, we're now almost ready to visit the locations. 

But first, let's look at a simple way to store the locations in one place.

# <hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 2.4</span>

We're almost ready to visit the locations in London! 

However, it's too cumbersome to lug around the 4 different lists we created in <span style="color:RoyalBlue">Exercise 2.3</span>.

<br>
**Create a single dictionary named <code style="color:steelblue">location_dict</code> for the locations in each borough.**
* Each key should be the name of the borough.
* Their values should be the lists of unique locations you created in <span style="color:RoyalBlue">Exercise 2.3</span>.

In [57]:
# Create location_dict
location_dict = {
    'Brent' : brent, 
    'Camden' :  camden, 
    'Redbridge' :  redbridge, 
    'Southwark' : southwark
}

print(location_dict)

{'Brent': ['Brent Park', 'Church End', 'Dollis Hill', 'Harlesden', 'Kensal Green', 'Kingsbury', 'Neasden', 'Preston', "Queen's Park", 'Tokyngton', 'Wembley', 'Wembley Park', 'Willesden', 'Park Royal', 'Kenton', 'Kilburn', 'Sudbury', 'Brondesbury', 'Alperton'], 'Camden': ['Bloomsbury', "King's Cross", 'Primrose Hill', 'Kentish Town', 'Holborn', 'Hampstead', 'Dartmouth Park', 'Highgate', 'West Hampstead', 'Camden Town', 'Fitzrovia', 'Frognal', 'Swiss Cottage', 'Chalk Farm', 'Belsize Park', 'Gospel Oak', 'St Giles', 'Somerstown', 'St Pancras'], 'Redbridge': ['Wanstead', 'Gants Hill', 'Goodmayes', 'Chadwell Heath', 'Snaresbrook', 'Redbridge', 'Ilford', 'Hainault', 'Loxford', 'Newbury Park', 'Aldborough Hatch', 'Barkingside', 'South Woodford', 'Seven Kings', 'Woodford', 'Woodford Green'], 'Southwark': ['Camberwell', 'Peckham', 'Bankside', 'Bermondsey', 'Dulwich', 'Nunhead', 'Elephant and Castle', 'Newington', 'Rotherhithe', 'Surrey Quays', 'Walworth', 'East Dulwich', 'Denmark Hill']}


Next, let's make sure the dictionary has the correct keys. 

<br>
**Run the cell below and check the output.**
* What do you think the code below is doing?
* You'll learn more about <code style="color:steelblue">for</code> loops in the next lesson.

In [56]:
# Run this cell
for borough in ['Brent', 'Camden', 'Redbridge', 'Southwark']:
    print( borough in location_dict )

True
True
True
True


<br>
Did you get the expected output? If not, check the answer key before moving on.

Suddenly, Q walks over to you and says...

> "Hmm... if you set up the dictionary correctly, you won't need the original lists anymore."

> "Please get rid of them."

<br>
**Run this next code cell to overwrite the original borough lists with <code style="color:steelblue">None</code>.**

In [58]:
# Run this cell
brent, camden, redbridge, southwark = None, None, None, None

By the way, <code style="color:steelblue">None</code> is its own object type in Python.

<br>
> *<span style="color:tomato; font-weight:bold">None</span> is an object that denotes emptiness.*

<br>
For example:

In [62]:
print( type(None) )
print(brent)

<class 'NoneType'>
None


Now, we want to split our visit to London into two trips: one for Inner London and one for Outer London.

<br>
**Add two new items to your dictionary:**
1. **Key:** <code style="color:steelblue">'Inner London'</code>... **Value:** All locations in <code style="color:steelblue">'Camden'</code> and <code style="color:steelblue">'Southwark'</code>.
2. **Key:** <code style="color:steelblue">'Outer London'</code>... **Value:** All locations in <code style="color:steelblue">'Brent'</code> and <code style="color:steelblue">'Redbridge'</code>.

Since you got rid of your original lists, you'll have to use the values you've already stored in your dictionary.

In [67]:
# Create a new key-value pair for 'Inner London'
location_dict['Inner London'] = location_dict['Camden'] + location_dict['Southwark'] 

# Create a new key-value pair for 'Outer London'
location_dict['Outer London'] = location_dict['Brent'] + location_dict['Redbridge'] 

print(location_dict['Inner London'])

['Bloomsbury', "King's Cross", 'Primrose Hill', 'Kentish Town', 'Holborn', 'Hampstead', 'Dartmouth Park', 'Highgate', 'West Hampstead', 'Camden Town', 'Fitzrovia', 'Frognal', 'Swiss Cottage', 'Chalk Farm', 'Belsize Park', 'Gospel Oak', 'St Giles', 'Somerstown', 'St Pancras', 'Camberwell', 'Peckham', 'Bankside', 'Bermondsey', 'Dulwich', 'Nunhead', 'Elephant and Castle', 'Newington', 'Rotherhithe', 'Surrey Quays', 'Walworth', 'East Dulwich', 'Denmark Hill']


Finally, let's just check that we have the right number of locations for each trip.
* You should have 32 for Inner London
* You should have 35 for Outer London

<br>
**Run the cell below and check that you get the expected output.**

In [69]:
print( len(location_dict['Inner London']) )
print( len(location_dict['Outer London']) )

32
35


AttributeError: 'dict' object has no attribute 'pkl'

If you don't have the right number of locations, doublecheck that you removed duplicates and that you're concatenating the correct lists. You can also check the answer key for the solution.

<br>
**Once you have the right number of locations, let's save this object so we can use it in the next lesson. Run the cell below.**
* We'll use a Python built-in package called <code style="color:steelblue">pickle</code> to do so.
* Pickle saves an entire object in a file on your computer.

In [70]:
import pickle

# Save object to disk
with open('location_dict.pkl', 'wb') as f:
    pickle.dump(location_dict, f)

<_io.BufferedWriter name='location_dict.pkl'>


<br>
***The mission continues in the next lesson...***

> *Now you have a dictionary of locations in London by borough.*

> *In the next lesson, we'll look through the locations and pick one to start with.*

<br>

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

<div style="text-align:center; margin: 40px 0 40px 0;">
[**Back to Contents**](#toc)
</div>

<br id="l3">
# Lesson 3: Flow and Functions

First, import the print function (if you've closed and reopened Jupyter since last time).

In [ ]:
from __future__ import print_function  # Compatability with Python 3
print('Print function ready to serve.')

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">
## <span style="color:RoyalBlue">Exercise 3.1</span>

***In the previous lesson...***

> *In the previous lesson, you created a location dictionary for the boroughs we're interested in visting.*

Now we're ready to pick a location to start with.

<br>
**First, let's import <code style="color:steelblue">location_dict</code> again using <code style="color:steelblue">pickle</code>. Run this cell.**

In [71]:
import pickle

# Read object from disk
with open('location_dict.pkl', 'rb') as f:
    location_dict = pickle.load(f)

Now we have the <code style="color:steelblue">location_dict</code> object again, but what if we forgot which keys are in the dictionary? 

<br>
**Print the keys in <code style="color:steelblue">location_dict</code>.**

In [72]:
# Print the keys in location_dict
print(location_dict.keys())

dict_keys(['Brent', 'Camden', 'Redbridge', 'Southwark', 'Inner London', 'Outer London'])


Ah, yes... 

Now, we need to choose between starting with <code style="color:steelblue">'Outer London'</code> or with <code style="color:steelblue">'Inner London'</code>. We should start with the list with more locations, so let's find which one it is.

<br>
**Write code, using <code style="color:steelblue">if</code> statements, that does the following:**
* **If** our Inner London list has more locations than our Outer London list, print the message:


<pre style="color:steelblue">I want to start in Inner London.</pre>


* **Else if** our Outer London list has more locations than our Inner London list, print the message:

    
<pre style="color:steelblue">I want to start in Outer London.</pre>


* **Else** (i.e. they have the same number of locations), print the message:


<pre style="color:steelblue">Either is fine. Flip a coin, Q.</pre>



In [73]:
# Write code here
if len(location_dict['Inner London']) > len(location_dict['Outer London']):
    print("I want to start in Inner London")
elif len(location_dict['Outer London']) > len(location_dict['Inner London']):
    print("I want to start in Outer London")
else :
    print("Either is fine. Flip a coin, Q.")


I want to start in Outer London


<hr style="border-color:royalblue;background-color:royalblue;height:1px;">
## <span style="color:RoyalBlue">Exercise 3.2</span>

Remember in <span style="color:RoyalBlue">Exercise 3.1</span>, we said that we wanted to start with the list with the most locations.
* We already knew that it would be either Inner London or Outer London because those lists are combinations of 2 of the others.
* However, what if we didn't know that?

<br>
**For each key in <code style="color:steelblue">location_dict</code>, print the number of locations in its list, like so:**

<pre style="color:steelblue">
Southwalk has 13 locations.
Redbridge has ...
</pre> 

* **Tip:** You can iterate through keys and values of a dictionary at the same time using <code style="color:steelblue">.items()</code>, like so:

<pre style="color:#bbb">
for <strong style="color:steelblue">key, value</strong> in location_dict<strong style="color:steelblue">.items()</strong>:
    <span style="color:dimgray"># code block</span>
</pre>      
        
* **Tip:** Remember, to insert multiple dynamic values into a string, you can just add more places to <code style="color:steelblue">.format()</code>, like so:

<pre style="color:#bbb">
'<strong style="color:steelblue">{}</strong> has <strong style="color:steelblue">{}</strong> locations.'.format(<strong style="color:steelblue">first_value, second_value</strong>)
</pre>

In [74]:
# For each key in location_dict, print the number of locations in its list
for key, value in location_dict.items():
    print("{} has {} locations.".format(key,len(value)),end='\n')


Brent has 19 locations.
Camden has 19 locations.
Redbridge has 16 locations.
Southwark has 13 locations.
Inner London has 32 locations.
Outer London has 35 locations.


Now, let's give each location in Outer London a first impression based on its name. 

<br>
**Combine <code style="color:steelblue">if</code> and <code style="color:steelblue">for</code> statements. For each location in Outer London...**
* **If** its name has <code style="color:steelblue">'Farm'</code>, <code style="color:steelblue">'Park'</code>, <code style="color:steelblue">'Hill'</code>, or <code style="color:steelblue">'Green'</code> in it, print:

<pre style="color:steelblue">{<strong>name</strong>} sounds pleasant.</pre>

* **Else If** its name has <code style="color:steelblue">'Royal'</code>, <code style="color:steelblue">'Queen'</code>, or <code style="color:steelblue">'King'</code> in it, print:

<pre style="color:steelblue">{<strong>name</strong>} sounds grand.</pre>
    
* If its name doesn't sound pleasant or grand, just ignore it.
* **Tip:** If you want to check if any word from a list is found in a string, you can use <code style="color:steelblue">any()</code>, like so:

<pre style="color:steelblue">
any( word in name for word in <strong>list_of_words</strong> )
</pre>

In [78]:
pleasant_sounding = ['Farm', 'Park', 'Hill', 'Green']
royal_sounding = ['Royal', 'Queen', 'King']

# Print first impression of each location in Outer London based on its names
for name in location_dict['Outer London'] :
    if any(word in name for word in pleasant_sounding) :
        print(name, 'sounds pleasant.')
    elif any(word in name for word in royal_sounding) :
        print(name, 'sounds grand.')


Brent Park sounds pleasant.
Dollis Hill sounds pleasant.
Kensal Green sounds pleasant.
Kingsbury sounds grand.
Queen's Park sounds pleasant.
Wembley Park sounds pleasant.
Park Royal sounds pleasant.
Gants Hill sounds pleasant.
Newbury Park sounds pleasant.
Seven Kings sounds grand.
Woodford Green sounds pleasant.


<hr style="border-color:royalblue;background-color:royalblue;height:1px;">
## <span style="color:RoyalBlue">Exercise 3.3</span>

**Using a list comprehension, create a new list called <code style="color:steelblue">pleasant_locations</code>.**
* It should contain locations in Outer London that are <code style="color:steelblue">pleasant_sounding</code>.
* Then print the list.
* **Tip:** To check if any word from a list is found in a string, you can use <code style="color:steelblue">any()</code>. See <span style="color:RoyalBlue">Exercise 4.2</span> if you need a reminder.

In [79]:
# Create pleasant_locations list using a list comprehension
pleasant_locations = [name for name in location_dict['Outer London'] if any(word in name for word in pleasant_sounding)]
# Print the pleasant-sounding locations
print(pleasant_locations)

['Brent Park', 'Dollis Hill', 'Kensal Green', "Queen's Park", 'Wembley Park', 'Park Royal', 'Gants Hill', 'Newbury Park', 'Woodford Green']


**Print the number pleasant-sounding locations we have.**

In [80]:
# Print number of pleasant-sounding locations
print(len(pleasant_locations))

9


<hr style="border-color:royalblue;background-color:royalblue;height:1px;">
## <span style="color:RoyalBlue">Exercise 3.4</span>

**Write a function called <code style="color:steelblue">filter_locations</code> that takes two arguments:**
1. <code style="color:steelblue">location_list</code>
2. <code style="color:steelblue">words_list</code>

The function should return the list of names in <code style="color:steelblue">location_list</code> that have any word in <code style="color:steelblue">words_list</code>.
* **Tip:** For help with the code block, see your answer in <span style="color:RoyalBlue">Exercise 4.3</span>.

In [81]:
# Code here
def filter_locations(location_list, words_list) :
    names = [name for name in location_list if any(word in name for word in words_list)]
    return names


Next, let's test that function. 

<br>
**Create a new <code style="color:steelblue">pleasant_locations</code> list using the function you just wrote.**
* Pass in the list of Outer London locations and the list of pleasant-sounding words.
* You should get the same locations that you got in <span style="color:RoyalBlue">Exercise 3.3</span>.
* Print the new list.

In [82]:
# Create pleasant_locations using filter_locations()
pleasant_locations = filter_locations(location_dict['Outer London'],pleasant_sounding)

# Print list of pleasant-sounding locations
print(pleasant_locations)


['Brent Park', 'Dollis Hill', 'Kensal Green', "Queen's Park", 'Wembley Park', 'Park Royal', 'Gants Hill', 'Newbury Park', 'Woodford Green']


**Next, let's use this handy function to create a <code style="color:steelblue">grand_locations</code> list for locations that sound grand.**
* Pass in the list of Outer London locations and the list of grand-sounding words.
* Print the new list and confirm the expected output

In [83]:
# Create grand_locations using filter_locations()
grand_locations = filter_locations(location_dict['Outer London'],royal_sounding)

# Print list of grand-sounding locations
print(grand_locations)


['Kingsbury', "Queen's Park", 'Park Royal', 'Seven Kings']


Awesome, are there any locations in both lists?

<br>
**Display the locations in both the <code style="color:steelblue">pleasant_locations</code> and <code style="color:steelblue">grand_locations</code> lists.**
* You can compare these lists manually because they are short, but try using <code style="color:steelblue">sets</code> and the <code style="color:steelblue">.intersection()</code> function (or <code style="color:steelblue">&</code> operator)!

In [86]:
# Display locations that sound pleasant and grand
pleasant_set = set(pleasant_locations)
grand_set = set(grand_locations)

print(pleasant_set.intersection(grand_set))


{"Queen's Park", 'Park Royal'}


Great, we'll start with these for our visit.

<br>
** *The mission continues in the next lesson...* **

> *In this lesson, we filtered our lists of locations down to two: Park Royal and Queen's Park.*

> *In the next lesson, we'll visit Park Royal and talk with some local businesses there.*

<br>
<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

<div style="text-align:center; margin: 40px 0 40px 0;">
[**Back to Contents**](#toc)
</div>

<br id="l4">
# Lesson 4: NumPy

First, import the print function (if you've closed and reopened Jupyter since last time).

In [ ]:
from __future__ import print_function  # Compatability with Python 3
print('Print function ready to serve.')

Then, let's import NumPy.

In [87]:
import numpy as np

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 4.1</span>

***In the previous lesson...***

> *In the previous lesson, we decided on Park Royal as our first stop in our training mission to London.*

Now we've arrived at the tube station!

<p style="text-align:center; margin-top:30px; margin-bottom:30px">
<img alt="Park Royal Tube Station" style="width: 320px; box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);"  src="https://elitedatascience.com/wp-content/uploads/2017/02/park-royal-tube.jpg">
<br>
<em>Park Royal Tube Station</em>
</p>

Park Royal is home to London's largest business park. 
* The business park supports around 1,700 businesses.
* Q hands you a manual with their names and locations. 
* *They are listed in order from smallest to largest* (this will be important later).

<br>
**First, let's create a NumPy array called <code style="color:steelblue">business_ids</code>.**
* The first business has ID <code>1</code>, the second one has ID <code>2</code> and so on.
* The array should have shape <code style="color:steelblue">(1700,)</code>.

In [91]:
# Create array of business_ids with values ranging from 1 to 1700
business_ids = np.array(range(1,1701))


**Next, print the shape of <code style="color:steelblue">business_ids</code> to confirm it has shape <code style="color:steelblue">(1700, )</code>.**

In [92]:
# Print shape of business_ids

print(business_ids.shape)

(1700,)


**Finally, print the last 10 business ID's to confirm the array is set up properly.**

In [94]:
# Print last 10 business ID's
print(business_ids[-10:])


[1691 1692 1693 1694 1695 1696 1697 1698 1699 1700]


<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 4.2a</span>

There are **1700** businesses in Park Royal's business park. However, you only have **10** hours to stay in Park Royal before you need to move on to the next location. Assuming you could only visit **5** businesses per hour, which businesses should you visit?

Well, 2 options immediately come to mind:
1. You could just visit the first 50 businesses.
2. You could randomly sample 50 businesses.

While these appear fine at first glance, there are potential flaws with both of these approaches. Remember, in <span style="color:royalblue">Exercise 4.1</span>, we learned that the businesses were listed in order from smallest to largest! Therefore...
1. Just visiting the first 50 would give us a biased sample of only the smallest businesses.
2. Visiting a random sample of 50 is better, but there's still a chance that our sample ends up biased from pure chance.

Instead, let's take a **stratified random sample** based on the size of the business.

### WTF is stratified random sampling?

This is a very important concept for practical machine learning. 

<br>
> *<span style="color:tomato; font-weight:bold">Stratified random sampling</span> is first grouping your observations based on a key variable, and then randomly sampling from those groups.* 

<br>

This will ensure your sample is **representative** of the broader dataset along that key variable.

<p style="text-align:center">
<img src="https://upload.wikimedia.org/wikipedia/commons/f/fa/Stratified_sampling.PNG" alt="Stratified Random Sampling" style="width: 280px;"/>
<br>
<em>Stratified Random Sampling</em>
</p>

For this exercise, it just means that...
* We'll start by splitting our businesses into 10 groups of 170 businesses each.
* The first group will have ID's from 1 to 170, the second will have ID's from 171 to 340, etc.
* Then we'll randomly select 5 businesses from each group of 170.
* Since the businesses are already ordered by size, this will ensure that small, medium, and big businesses are all represented in our sample.

<br>
**First, reshape the 1-dimensional <code style="color:steelblue">business_ids</code> array into a new 2-dimensional <code style="color:steelblue">id_matrix</code> array.**
* <code style="color:steelblue">id_matrix</code> should have **10 columns**... one for each group of businesses.
* How many rows should it have?
* What does the number of rows represent?

In [95]:
# Create id_matrix by reshaping business_ids to have 10 columns
id_matrix = business_ids.reshape(170,10)

# Print shape
print(id_matrix.shape)


(170, 10)


Great, now we have a matrix with 10 columns representing 10 groups of businesses. But remember, our goal is to stratify our sample by size of the business, and because our businesses are ordered by size, the first group should be 1 to 170, the second group should be 171 to 340, and so on.

<br>
**Print the first column (group) of <code style="color:steelblue">id_matrix</code>.**
* Does it contain businesses 1 to 170?

In [97]:
# Print first column of id_matrix
print(id_matrix[:,0])


[   1   11   21   31   41   51   61   71   81   91  101  111  121  131  141
  151  161  171  181  191  201  211  221  231  241  251  261  271  281  291
  301  311  321  331  341  351  361  371  381  391  401  411  421  431  441
  451  461  471  481  491  501  511  521  531  541  551  561  571  581  591
  601  611  621  631  641  651  661  671  681  691  701  711  721  731  741
  751  761  771  781  791  801  811  821  831  841  851  861  871  881  891
  901  911  921  931  941  951  961  971  981  991 1001 1011 1021 1031 1041
 1051 1061 1071 1081 1091 1101 1111 1121 1131 1141 1151 1161 1171 1181 1191
 1201 1211 1221 1231 1241 1251 1261 1271 1281 1291 1301 1311 1321 1331 1341
 1351 1361 1371 1381 1391 1401 1411 1421 1431 1441 1451 1461 1471 1481 1491
 1501 1511 1521 1531 1541 1551 1561 1571 1581 1591 1601 1611 1621 1631 1641
 1651 1661 1671 1681 1691]


Crap, that's not what we wanted. Ah, but we should have seen this coming!

Remember, when you <code style="color:steelblue">.reshape</code> an array, the new array keeps the order of the elements. Continue reading to see how we can solve this.

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 4.2b</span>

Let's try building that <code style="color:steelblue">id_matrix</code> again. 

<br>
**This time, reshape it to have 10 rows, each representing 1 group of businesses.**
* The first row should contain businesses 1 to 170.

In [98]:
# Reshape business_ids to have 10 rows, with 170 businesses each
id_matrix = business_ids.reshape(10,170)


**Now, print the first row of <code style="color:steelblue">id_matrix</code>.**

In [99]:
# Print the first row of id_matrix
print(id_matrix[0,:])


[  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
 163 164 165 166 167 168 169 170]


You should see the businesses with ID's from **1 to 170**.

Now, what if we still want our groups to be grouped by column, instead of by row? Let's *flip our rows and columns* now that the rows have the correct groups.

<br>
**Overwrite <code style="color:steelblue">id_matrix</code> with its transposed version.**
* After you do so, print the first column in the new <code style="color:steelblue">id_matrix</code> to confirm the group is correct.
* Also print the shape of the new <code style="color:steelblue">id_matrix</code> to confirm your new dimensions are correct.

In [103]:
# Overwrite id_matrix with flipped version
id_matrix = id_matrix.transpose()

# Print first column
print(id_matrix[:,0])

# Print shape of new id_matrix
print(id_matrix.shape)


[  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
 163 164 165 166 167 168 169 170]
(170, 10)


<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 4.3</span>

Let's take another look at our <code style="color:steelblue">id_matrix</code>. 

* In <span style="color:royalblue">Exercise 4.2b</span> we displayed the first group (column) and confirmed that the ID's were from **1 to 170**.
* Now let's confirm the rest of the array is correct by finding the <code style="color:steelblue">np.min()</code> and <code style="color:steelblue">np.max()</code> of each group.

<br>
**First, create and print an object called <code style="color:steelblue">group_min</code> with the minimum ID of each group (column) of <code style="color:steelblue">id_matrix</code>.**

In [105]:
# Create group_min
group_min = np.min(id_matrix,axis=0) 


# Print group_min
print(group_min)


[   1  171  341  511  681  851 1021 1191 1361 1531]


**Next, create and print an object called <code style="color:steelblue">group_max</code> with the maximum ID of each group (column) of <code style="color:steelblue">id_matrix</code>.**

In [106]:
# Create group_max
group_max = np.max(id_matrix,axis=0) 

# Print group_max
print(group_max)


[ 170  340  510  680  850 1020 1190 1360 1530 1700]


**Finally, subtract <code style="color:steelblue">group_min</code> from <code style="color:steelblue">group_max</code> to confirm that each of the 10 groups has a <span style="tomato">range</span> of 170.**
* Remember to add 1 to the difference between max and min because the ends are inclusive 
    * i.e. $170 - 1 = 169$

In [107]:
# Print range of each group
print(group_max - group_min + 1)


[170 170 170 170 170 170 170 170 170 170]


<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 4.4</span>

Great job so far.

We are now ready to select 5 businesses from each group in <code style="color:steelblue">id_matrix</code>.

<br>
**Write a loop that chooses 5 businesses from column of <code style="color:steelblue">id_matrix</code>.**
* Set the random seed to 9001.
* Format the output like in the <span style="color:royalblue">Expected output</span> cell below.
* Should we sample with replacement or without replacement? **Hint:** What happens if a business is selected twice?
* Bonus points for writing the loop in a way that works with any number of columns.

In [114]:
# Seed random seed
np.random.seed(9001)

# Print selected businesses from each group
for x in range(0,id_matrix.shape[1]) :
    print( np.random.choice(id_matrix[:,x], 5, replace = False ))



[  7  37 124  41  17]
[302 261 257 323 234]
[464 486 463 440 474]
[645 582 666 631 553]
[699 738 705 760 792]
[ 995 1000  909  869  899]
[1042 1058 1172 1122 1153]
[1304 1275 1343 1344 1236]
[1401 1417 1383 1387 1474]
[1642 1687 1545 1644 1549]


Awesome, now we're ready to start visiting the businesses... Let's start with Group 1 and the business with ID 7. 

After checking our manual, we find out that this business happens to be **Casey's Flower Shop**!

<p style="text-align:center">
<img src="https://elitedatascience.com/wp-content/uploads/2017/02/caseys-flower-shop.jpg" alt="Caseys Flower Shop" style="width: 280px; box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19); margin-top:20px">
<br>
<em>The <s>cheesy</s> welcoming banner outside Casey's door.</em>
</p>

WTF does Casey's Flower Shop have to do with anything? Find out <s>next time on Dragon Ball Z</s> in the next lesson!

<br>
** *The mission continues in the next lesson...* **

> *In this lesson, you narrowed the list of businesses down through stratified random sampling.*

> *In the next lesson, you'll visit Casey's Flower Shop and offer help using your data analysis skills!*

<br>
<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

<div style="text-align:center; margin: 40px 0 40px 0;">
[**Back to Contents**](#toc)
</div>

<br id="l5">
# Lesson 5: Pandas

First, import the print function (if you've closed and reopened Jupyter since last time).

In [ ]:
from __future__ import print_function  # Compatability with Python 3
print('Print function ready to serve.')

Then, let's import Pandas.

In [7]:
import pandas as pd

Read Iris dataset:

In [16]:
# Read the iris dataset from a CSV file
df = pd.read_csv('project_files/iris.csv')

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 5.1</span>

***In the previous lesson...***

> *In the previous lesson, we discovered that **Casey's Flower Shop** is the first local business we need to visit in Park Royal.*


When arrive, we find out that the owner, Casey, urgently needs our help!
> * They just received a new shipment of **Iris flowers**, but they've never stocked these flowers before.
* Casey asks us to share what we know about these new flowers.
* Luckily, we have the Iris dataset to learn more about them ourselves!

In Lesson 4: NumPy, we also introduced the idea of using a **toy problem** to reduce the size of a problem and break down the concepts.

While 150 observations is not exactly "big data," it's still too large to fit on our screen at once. So let's use another toy problem to practice the concepts in this lesson.

<br>
**First, create a new DataFrame called <code style="color:steelblue">toy_df</code>. It should contain the first 5 rows plus the last 5 rows from our original Iris dataset.**
* **Tip:** You already have a <code style="color:steelblue">.head()</code>, but what about a <code style="color:steelblue">.tail()</code>?
* **Tip:** <code style="color:steelblue">pd.concat()</code> is your [friend](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html).

In [34]:
# Create toy_df
toy_df = pd.concat([df.head(),df.tail()])


**Next, display <code style="color:steelblue">toy_df</code>.** 
* After all, it will only be 10 rows.
* In <code style="color:steelblue">toy_df</code>, you should have data from 2 different species of flower. Which are they?

In [35]:
# Display toy_df

toy_df

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

You should have 'setosa' and 'virginica' flowers.

**Next, display a summary table for <code style="color:steelblue">toy_df</code>.**
* It should have the mean, standard deviation, and quartiles for each of the columns

In [20]:
# Describe toy_df
toy_df.describe()



sepal_length  sepal_width  petal_length  petal_width
count     10.000000    10.000000     10.000000    10.000000
mean       5.590000     3.130000      3.290000     1.130000
std        0.807534     0.316403      1.995244     0.992248
min        4.600000     2.500000      1.300000     0.200000
25%        4.925000     3.000000      1.400000     0.200000
50%        5.500000     3.050000      3.250000     1.000000
75%        6.275000     3.350000      5.175000     1.975000
max        6.700000     3.600000      5.400000     2.300000

Since <code style="color:steelblue">toy_df</code> is only 10 rows, you can manually check the **mins** and **maxes**. Are they correct?

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 5.2</span>

Elementwise operations are very useful in machine learning, especially for feature engineering.

<br>

> *<span style="color:tomato; font-weight:bold">Feature engineering</span> is the process of creating new features (model input variables) from existing ones.*

<br>

We'll cover this topic in much more detail in Projects 2, 3, and 4, but let's first use our <code style="color:steelblue">toy_df</code> to illustrate the concept.

In the Iris dataset, we have petal width and length, but what if we wanted to know petal area? Well, we can create a new <code style="color:steelblue">petal_area</code> feature (yes, the petals are not perfect rectangles, but that's fine).

<br>
**First, display the two columns of <code style="color:steelblue">petal_width</code> and <code style="color:steelblue">petal_length</code> in <code style="color:steelblue">toy_df</code>.**
* **Tip:** You can index a DataFrame using a list of column names too, like so:


<pre style="color:steelblue">df[['column_1', 'column_2']]</pre>

In [36]:
# Display petal_width and petal_length

toy_df[['petal_width','petal_length']]

petal_width  petal_length
0            0.2           1.4
1            0.2           1.4
2            0.2           1.3
3            0.2           1.5
4            0.2           1.4
145          2.3           5.2
146          1.9           5.0
147          2.0           5.2
148          2.3           5.4
149          1.8           5.1

**Next, create a new <code style="color:steelblue">petal_area</code> feature in <code style="color:steelblue">toy_df</code>.**
* Multiply the <code style="color:steelblue">petal_width</code> column by the <code style="color:steelblue">petal_length</code> column.
* Display <code style="color:steelblue">toy_df</code> after creating the new feature.
* Are the values for <code style="color:steelblue">petal_area</code> correct? Manually spot check a few of them just to make sure.

In [37]:
# Create a new petal_area column
toy_df['petal_area']=toy_df.petal_width*toy_df.petal_length

# Display toy_df
toy_df['petal_area']


0       0.28
1       0.28
2       0.26
3       0.30
4       0.28
145    11.96
146     9.50
147    10.40
148    12.42
149     9.18
Name: petal_area, dtype: float64

**Finally, what do we now know about Iris flowers?**

By creating a <code style="color:steelblue">petal_area</code> feature, it's now much easier to see that virginica flowers have significantly larger petals than setosa flowers do!

Often, by creating new features, you can learn more about the data (and improve your machine learning models).

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 5.3</span>

Again, we'll use the <code style="color:steelblue">toy_df</code> to really drive home these concepts.

Let's say we wanted to display observations where <code style="color:steelblue">petal_area > 10</code> and <code style="color:steelblue">sepal_width > 3</code>. How could we do so?

<br>
**First, display <code style="color:steelblue">toy_df</code> again just to have it in front of you.**

In [38]:
# Display toy_df
toy_df


sepal_length  sepal_width  petal_length  petal_width    species  \
0             5.1          3.5           1.4          0.2     setosa   
1             4.9          3.0           1.4          0.2     setosa   
2             4.7          3.2           1.3          0.2     setosa   
3             4.6          3.1           1.5          0.2     setosa   
4             5.0          3.6           1.4          0.2     setosa   
145           6.7          3.0           5.2          2.3  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
148           6.2          3.4           5.4          2.3  virginica   
149           5.9          3.0           5.1          1.8  virginica   

     petal_area  
0          0.28  
1          0.28  
2          0.26  
3          0.30  
4          0.28  
145       11.96  
146        9.50  
147       10.40  
148       12.42  
149        9.18

**Take a look at the DataFrame and manually count the number that satisfy our conditions.**
* How many observations have <code style="color:steelblue">petal_area > 10</code>?
* How many observations have <code style="color:steelblue">sepal_width > 3</code>?
* How many satisfy both conditions?

Great. Now we'll see what's going on under the hood when we use our boolean masks.

<br>
**Create a boolean mask for <code style="color:steelblue">petal_area > 10</code>.**
* Name it <code style="color:steelblue">petal_area_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [39]:
# Mask for petal_area > 10
petal_area_mask = toy_df.petal_area > 10

# Display petal_area_mask
petal_area_mask


0      False
1      False
2      False
3      False
4      False
145     True
146    False
147     True
148     True
149    False
Name: petal_area, dtype: bool

**Next, create a boolean mask for <code style="color:steelblue">sepal_width > 3</code>.**
* Name it <code style="color:steelblue">sepal_width_mask</code>.
* Display the mask after you create it.
* Does the result make sense?

In [40]:
# Mask for sepal_width > 3
sepal_width_mask = toy_df.sepal_width > 3

# Display sepal_width_mask
sepal_width_mask


0       True
1      False
2       True
3       True
4       True
145    False
146    False
147    False
148     True
149    False
Name: sepal_width, dtype: bool

**Next, display the two masks combined using the <code style="color:steelblue">&</code> operator.**
* Note how their combination results in another boolean mask!

In [41]:
# Display both masks, combined
petal_area_mask & sepal_width_mask


0      False
1      False
2      False
3      False
4      False
145    False
146    False
147    False
148     True
149    False
dtype: bool

**Finally, select the observations from <code style="color:steelblue">toy_df</code> where both conditions are met.**

In [42]:
# Index with both masks
toy_df[petal_area_mask & sepal_width_mask]


sepal_length  sepal_width  petal_length  petal_width    species  \
148           6.2          3.4           5.4          2.3  virginica   

     petal_area  
148       12.42

<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

## <span style="color:RoyalBlue">Exercise 5.4</span>

Now, armed with the power of **groupby**, we're almost ready to return to Casey and share what we've learned about Iris flowers! But before we do, let's just bring back our <code style="color:steelblue">toy_df</code> for one last hoorah, just to make sure we know what's going on under the hood.

<br>
Let's calculate the median <code style="color:steelblue">petal_area</code> for each species. 
* Since <code style="color:steelblue">toy_df</code> is small, we can do this manually as well and check to make sure the values are correct.

<br>
**First, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the virginica flowers in our <code style="steelblue">toy_df</code>.**
* Display all observations of the virginica species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.
* **Tip:** Check out the <code style="color:steelblue">.sort_values()</code> function ([documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)).

In [51]:
# Display all 'virginica' species, sorted by petal_area
species_mask = toy_df.species=='virginica'
toy_df[species_mask].sort_values(by='petal_area')


sepal_length  sepal_width  petal_length  petal_width    species  \
149           5.9          3.0           5.1          1.8  virginica   
146           6.3          2.5           5.0          1.9  virginica   
147           6.5          3.0           5.2          2.0  virginica   
145           6.7          3.0           5.2          2.3  virginica   
148           6.2          3.4           5.4          2.3  virginica   

     petal_area  
149        9.18  
146        9.50  
147       10.40  
145       11.96  
148       12.42

Based on the output above, what's median <code style="color:steelblue">petal_area</code> for the virginica species?

<br>
**Next, let's manually calculate the median <code style="color:steelblue">petal_area</code> for the setosa flowers in our <code style="steelblue">toy_df</code>.**
* Display all observations of the setosa species.
* Sort them by <code style="color:steelblue">petal_area</code> in ascending order.

In [57]:
# Display all 'setosa' species
toy_df[toy_df.species=='setosa'].sort_values(by='petal_area')


sepal_length  sepal_width  petal_length  petal_width species  petal_area
2           4.7          3.2           1.3          0.2  setosa        0.26
0           5.1          3.5           1.4          0.2  setosa        0.28
1           4.9          3.0           1.4          0.2  setosa        0.28
4           5.0          3.6           1.4          0.2  setosa        0.28
3           4.6          3.1           1.5          0.2  setosa        0.30

Based on the output above, what's median <code style="color:steelblue">petal_area</code> for the setosa species?

<br>
**Finally let's calculate the median values using a <code style="color:steelblue">.groupby()</code>.**
* Do you get the same result?

In [59]:
# Median petal_area in toy_df
toy_df.groupby('species').median()

sepal_length  sepal_width  petal_length  petal_width  petal_area
species                                                                    
setosa              4.9          3.2           1.4          0.2        0.28
virginica           6.3          3.0           5.2          2.0       10.40

** *Congratulations... You've completed the training mission!* **

> *In this lesson, you explored the Iris dataset using Pandas.*

> *After arming yourself with Pandas, you've successfully completed the rest of the training mission!*

<br>
<hr style="border-color:royalblue;background-color:royalblue;height:1px;">

<div style="text-align:center; margin: 40px 0 40px 0;">
[**Back to Contents**](#toc)
</div>